File: src/reference_id.cpp
// src/reference_id.cpp — validate and store stable semantic reference identifiers.
#include "reference_id.hpp"

#include <stdexcept>

namespace
{
auto is_ascii_letter(const char character) noexcept -> bool
{
    return (character >= 'a' && character <= 'z') || (character >= 'A' && character <= 'Z');
}

auto is_ascii_digit(const char character) noexcept -> bool
{
    return character >= '0' && character <= '9';
}

auto is_valid_reference_character(const char character) noexcept -> bool
{
    return is_ascii_letter(character) || is_ascii_digit(character) || character == '-'
           || character == '_' || character == '.' || character == ':';
}
}  // namespace

namespace dans::document
{
ReferenceId::ReferenceId(const std::string_view value) : value_{value}
{
    if (value.empty())
    {
        throw std::invalid_argument{"A reference ID must not be empty"};
    }
    if (!is_ascii_letter(value.front()))
    {
        throw std::invalid_argument{"A reference ID must begin with an ASCII letter"};
    }
    for (const char character : value)
    {
        if (!is_valid_reference_character(character))
        {
            throw std::invalid_argument{
                "A reference ID may contain only ASCII letters, digits, '-', '_', '.', and ':'"
            };
        }
    }
}

auto ReferenceId::value() const noexcept -> std::string_view
{
    return value_;
}
}  // namespace dans::document

File: builder/src/model/referenceId.ts
// Stable semantic IDs shared by reference-producing and reference-consuming plugins.
export const referenceIdPattern = /^[A-Za-z][A-Za-z0-9_.:-]*$/u;

export function validateOptionalReferenceId(
  referenceId: string | null,
  context: string,
): void {
  if (referenceId !== null && !referenceIdPattern.test(referenceId)) {
    throw new Error(
      `${context} must begin with an ASCII letter and contain only letters, digits, '-', '_', '.', and ':'`,
    );
  }
}

export function decodeOptionalReferenceId(
  value: unknown,
  context: string,
): string | null {
  // Missing fields are accepted as null for payloads written before targets
  // were added to figures, equations, and listings.
  if (value === null || value === undefined) {
    return null;
  }
  if (typeof value !== "string") {
    throw new Error(`${context} must be a string or null`);
  }
  validateOptionalReferenceId(value, context);
  return value;
}

export function requireReferenceId(value: string, context: string): string {
  validateOptionalReferenceId(value, context);
  if (value.length === 0) {
    throw new Error(`${context} cannot be empty`);
  }
  return value;
}
